README.md

```diff
@@ -629,6 +629,274 @@ Can you try updating the rendering function to also show `entry.content` and
+  | Loading /* A request is currently taking place to fetch the feed. */
+  | InvalidUsername /* The entered username is not a valid GitHub ID. */
+  | Loaded(result(Feed.feed, string)); /* A request has finished, its result is contained inside the variant value itself. */
+
+type state = {
+  username: string,
+  step,
+};
+```
+
+After this, we will replace the two state hooks that were using for `data` and
+`username` with a single one that handles this new state:
+
+```reasonml
+  let (state, setState) =
+    React.useState(() => {username: "jchavarri", step: Idle});
+```
+
+Let's modify `fetchFeed` to take a value of type `Username.t` rather than
+`string`, so we make sure there are no requests being done with invalid
+usernames. This essentially consists on replaing `username` with
+`Username.toString(username)`, in order to convert back to string when building
+the URL.
+
+Now, let's start thinking which parts of our component will modify the state:
+- When we call `fetchFeed`, we have to set `step` to `Loading` at the beginning
+  of the function (`setState(state => {...state, step: Loading});`), and
+  `Loaded` when the request has finished (`setState(state => {...state, step:
+  Loaded(data)})`)
+- When the input value changes, we will go back to step `Idle`:
+```reasonml
+  onChange={event => {
+    setState(_ =>
+      {username: event->React.Event.Form.target##value, step: Idle}
+    )
+  }}
+```
+- We validate the username on `Enter` to ensure only valid requests are sent. If
+  the name is valid, we can just call `fetchFeed`, but if it's invalid, we will
+  set the step to `InvalidUsername`:
+```reasonml
+  onKeyDown={event => {
+    let enterKey = 13;
+    if (React.Event.Keyboard.keyCode(event) == enterKey) {
+      switch (Username.make(state.username)) {
+      | Ok(username) => fetchFeed(username)
+      | Error () =>
+        setState(state => {...state, step: InvalidUsername})
+      };
+    };
+  }}
+```
+
+To continue, we have to adapt the effect to check the username is valid:
+
+```reasonml
+  React.useEffect0(() => {
+    switch (Username.make(state.username)) {
+    | Ok(username) => fetchFeed(username)
+    | Error () =>
+      Js.Exn.raiseError(
+        "Initial username passed to React.useState is invalid",
+      )
+    };
+    None;
+  });
+```
+
+Finally, we can modify the `switch` in the JSX code to use the new state, so
+instead of checking `data` we check for `state.step`:
+
+```reasonml
+  {switch (state.step) {
+    | InvalidUsername => <div> {React.string("Invalid username")} </div>
+    | Idle =>
+      <div>
+        {React.string(
+          "Press the \"Enter\" key to confirm the username selection.",
+        )}
+      </div>
+    | Loading => <div> {React.string("Loading...")} </div>
+    | Loaded(Error(msg)) => <div> {React.string(msg)} </div>
+    | Loaded(Ok(feed)) => ...
+  }}
+```
```

Comment:
After following along, there's one thing that was still not compiling, though it's pretty minor and not hard to fix: Changing `input`'s `value` prop from `value=username` to `value={state.username}`.

Reply:
https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/6d2eda65458312c280fbbc53c94949d4d37b30e7